src/engine/types/mod.rs

```diff
@@ -153,58 +154,44 @@ impl<'a> TryFrom<&'a str> for ReportType {
 
 #[derive(Clone, Debug)]
 pub struct BlockDevPath {
-    /// Path to the physical device.
-    physical_path: PathBuf,
-    /// Optional tuple for encrypted devices. The first path is the internal
-    /// path of the logical device and the second is the canonicalized path
-    /// of the logical device.
-    logical_path: Option<(PathBuf, PathBuf)>,
+    /// Path to the device represented by this data structure.
+    path: PathBuf,
+    /// Reference to the path of the child device of this device.
+    child_paths: Vec<Arc<BlockDevPath>>,
 }
 
 impl BlockDevPath {
-    /// Path for a device that only has a physical path.
-    pub fn physical_device_path(physical: &Path) -> BlockDevPath {
-        BlockDevPath {
-            physical_path: physical.to_owned(),
-            logical_path: None,
-        }
+    /// Create a new node in the graph representing a device with no children.
+    pub fn new(path: PathBuf) -> Arc<BlockDevPath> {
+        Arc::new(BlockDevPath {
+            path,
+            child_paths: vec![],
+        })
     }
 
-    /// Path for a device that has both an associated physical path and a mapped
-    /// logical path.
-    pub fn mapped_device_path(physical: &Path, logical: &Path) -> Result<BlockDevPath, io::Error> {
-        let logical_path = Some((logical.to_owned(), logical.canonicalize()?));
-        Ok(BlockDevPath {
-            physical_path: physical.to_owned(),
-            logical_path,
+    /// Create a new node in the graph representing the devices and their children.
+    pub fn new_with_children<I>(path: PathBuf, child_paths: I) -> Arc<BlockDevPath>
```

Comment:
I would normally agree, but in this case, `BlockDevPath`s really can't be used outside of the context of an `Arc` because of the graph data structure. I am happy to rename these methods to make it more clear, but I don't see much of a difference between this and returning a `Result<BlockDevPath>` from a new function for example. Both are simply wrapping the core type in another outer type.

Reply:
Yeah, we shouldn't be returning ```Result<T>``` from ```new``` functions. clippy rightly complains, but that's in too deep to get out easily, so we left it.

Reply:
Okay, is renaming adequate?

Reply:
Yes, rename will do it.

Reply:
Resolved.